feat: 전체 커밋 랭킹을 가져온다

hyojaekim opened the pull request.

hyojaekim committed:
feat: Github 커밋 랭킹 API 모든 접근 허용

hyojaekim committed:
feat: 날짜에 해당하는 커밋 랭킹 가져오기

hyojaekim committed:
feat: 전체 커밋 랭킹 가져오는 API

hyojaekim committed:
feat: 전체 커밋 랭킹을 가져온다(front)

hyojaekim committed:
fix: 나의 랭킹 정보 가져오기(front)

hyojaekim committed:
refactor: 오늘 나의 랭킹 정보가 없으면 예외를 던지도록 수정

hyojaekim committed:
refactor: 오늘 나의 랭킹 정보가 없으면 400 에러가 발생한다

hyojaekim committed:
test: 테스트 결과 확인

- 로그인 중인 유저의 커밋 랭킹 결과 확인
- 전체 커밋 랭킹 결과 확인

hyojaekim committed:
feat: 나의 랭킹이 없으면 해당 영역을 보여주지 않는다

hyojaekim committed:
feat: 커밋 랭킹 캐시 적용

hyojaekim committed:
refactor: 전체 커밋 랭킹 높이 조절

hyojaekim committed:
fix: 50위까지 조회할 수 있도록 수정

hyojaekim committed:
feat: 시작 랭크부터 10개씩 가져온다

- 만약에 시작 랭크부터 10개 이하(n개)라면, n개를 가져온다
- 시작하는 랭크가 0위 또는 51위 이상이면 예외가 발생한다
- 시작하는 랭크가 1, 11, 21, 31, 41가 아니라면 예외가 발생한다

hyojaekim committed:
refactor: 커밋 랭킹 가져오는 API 응답 결과 수정

- List<UserCommitRankDetailResponseDto> -> TotalCommitRankResponseDto
- 전체 랭킹을 가져왔는지 확인하는 필드 추가(isMaxRank)

hyojaekim committed:
feat: 커밋 랭킹 범위 넘어가는 예외 처리(400)

- 테스트 추가

hyojaekim committed:
feat: 스크롤마다 10개씩 랭킹을 가져온다(프론트)

hyojaekim committed:
chore: package.json, package-lock.json 수정

oeeen commented on woowacrew-web/src/main/java/woowacrew/github/service/GithubCommitSchedulerService.java:
```diff
@@ -16,6 +17,7 @@ public GithubCommitSchedulerService(GithubCommitService githubCommitService) {
     }
 
     @Scheduled(cron = "0 0 9 * * *")
+    @CacheEvict(value = "totalCommitRank")
```
👍

oeeen commented on woowacrew-web/src/main/java/woowacrew/security/AbstractSecurityConfig.java:
```diff
@@ -72,6 +72,7 @@ protected void configure(HttpSecurity http) throws Exception {
                 .access("hasRole('ROLE_ADMIN') or hasIpAddress('" + jenkinsIpv4Address + "') or hasIpAddress('" + jenkinsIpv6Address + "')")
                 .antMatchers("/h2-console", "/h2-console**", "/h2-console/", "/h2-console/**").permitAll()
                 .antMatchers("/", "/error", "/login", "/login/**", "/search", "/search/**", "/docs/**", "/api/slack", "/api/slack/birthday-message").permitAll()
+                .antMatchers("/api/github/commit/rank").permitAll()
```
role_precourse에는 api open 안해줘도 될 것 같지만 일단은 괜찮은 것 같습니다.

oeeen commented on woowacrew-web/src/main/java/woowacrew/github/service/GithubCommitService.java:
```diff
@@ -29,7 +38,16 @@ public void save(GithubCommitRequestDto requestDto) {
 
     private LocalDate createDate(GithubCommitRequestDto requestDto) {
-        return LocalDate.of(requestDto.getYear(), requestDto.getMonth(), 1);
+        return createDate(requestDto.getYear(), requestDto.getMonth());
```
단순 의견입니다.

이 코드를 처음 봤을 때 GithubCommitRequestDto가 뭐길래 여기서 년,월을 꺼내서 날짜를 만들지? 했습니다. GithubCommitRequestDto는 년, 월로 이루어져 있었구요. dto이긴 하지만 네이밍(githubCommitRequest)에서 내부의 데이터들을 표현하고 있지는 못하는 것 같아요.

hyojaekim replied:
다시 보니 Dto의 네이밍이 적절하지 못하다고 생각했습니다! 
고민해서 반영하도록 하겠습니다 👍

oeeen commented on woowacrew-web/src/main/java/woowacrew/github/service/GithubCommitService.java:
```diff
@@ -40,4 +58,63 @@ public UserCommitRankDetailResponseDto getLoginUserCommitRank(UserContext userCo
+        }
+    }
+
+    public int getEndRank(int startRank, int maxRank) {
```
이 메서드는 이 클래스 내부에서만 사용하는데 public 인데, 테스트 때문에 public으로 푸셨나요?

hyojaekim replied:
외부에서 사용하다가 내부에서 사용하도록 수정했는데
private로 변경하겠습니다!

oeeen commented on woowacrew-web/src/main/java/woowacrew/github/service/GithubCommitService.java:
```diff
@@ -40,4 +58,63 @@ public UserCommitRankDetailResponseDto getLoginUserCommitRank(UserContext userCo
+        return endRank;
+    }
+
+    private void checkMaxRank(int maxRank) {
```
단순 의견입니다.

Rank를 validation 하는 부분이 생각보다 많이 들어있는데, 혹시 따로 클래스로 빼는 건 어떨까요? 예전에 LottoNumber같은 wrapping한 클래스를 만드는 걸 생각해봤습니다.

hyojaekim replied:
적절한 클래스가 수행하도록 수정하겠습니다.
좋은 피드백 감사합니다!! 👍

oeeen reviewed (approved):
고생했습니다 효오! 👍 

리뷰 요청한지 2주 넘었는데 이제야 리뷰하네요.. 그것도 자세히는 못봤어요.. 미안해요.. 
개인적으로 코드 보면서 든 생각, 의견들 남겨봤어요.

의견만 드리고 pr은 approve 하겠습니다~ 💯

hyojaekim committed:
refactor: 접근 제어자 변경(endRank)

- public -> private

hyojaekim committed:
style: 클래스명 변경

- GithubCommitRequestDto -> ThisMonthCommitRankRequestDto

hyojaekim committed:
style: 메서드명 변경

- createDate -> getFirstDateByYearAndMonth
- createDate -> getFirstDateOfThisMonth

hyojaekim committed:
refactor: 랭킹 유효성 체크 클래스 생성

- 랭킹에 대한 유효성 체크는 GithubCommitRank 클래스가 하도록 수정
